File: examples/algorithm-template.py
"""
Template for algorithm implementations

Copy this template when creating new algorithm files.
"""

"""
Problem: [Problem Name Here]

Description:
[Provide a clear, concise description of the problem you're solving]

Algorithm:
[Explain the approach and strategy used to solve the problem]

Time Complexity: O(?)
Space Complexity: O(?)

Example:
Input: [example input]
Output: [expected output]

Edge Cases:
- [List important edge cases to consider]
- [Empty input, single element, duplicates, etc.]
"""

def solution_function(parameters):
    """
    Brief description of what this function does
    
    Args:
        parameters: Description of input parameters
        
    Returns:
        Description of what the function returns
        
    Raises:
        ValueError: Description of when this might be raised
    """
    # Step 1: Handle edge cases
    if not parameters:
        return None  # or appropriate default
    
    # Step 2: Initialize variables
    # [Initialize any needed variables]
    
    # Step 3: Main algorithm logic
    # [Implement the core algorithm]
    
    # Step 4: Return result
    return result

# Alternative solutions (if applicable)
def solution_function_optimized(parameters):
    """
    Optimized version of the algorithm
    
    Time Complexity: O(?)
    Space Complexity: O(?)
    """
    pass

# Test cases
def test_solution():
    """Test cases for the algorithm"""
    
    # Test case 1: Normal case
    assert solution_function([1, 2, 3]) == expected_result
    
    # Test case 2: Edge case - empty input
    assert solution_function([]) == expected_empty_result
    
    # Test case 3: Edge case - single element
    assert solution_function([1]) == expected_single_result
    
    # Test case 4: Edge case - duplicates
    assert solution_function([1, 1, 1]) == expected_duplicate_result
    
    print("✅ All test cases passed!")

if __name__ == "__main__":
    test_solution()

File: examples/algorithm-template.js
/**
 * Template for Node.js algorithm implementations
 *
 * Copy this template when creating new algorithm files.
 */

/**
 * Problem: [Problem Name Here]
 *
 * Description:
 * [Provide a clear, concise description of the problem you're solving]
 *
 * Algorithm:
 * [Explain the approach and strategy used to solve the problem]
 *
 * Time Complexity: O(?)
 * Space Complexity: O(?)
 *
 * Example:
 * Input: [example input]
 * Output: [expected output]
 *
 * Edge Cases:
 * - [List important edge cases to consider]
 * - [Empty input, single element, duplicates, etc.]
 */

/**
 * Brief description of what this function does
 *
 * @param {any} parameters - Description of input parameters
 * @returns {any} Description of what the function returns
 * @throws {Error} Description of when this might throw an error
 */
function solutionFunction(parameters) {
  // Step 1: Handle edge cases
  if (!parameters) {
    return null; // or appropriate default
  }

  // Step 2: Initialize variables
  // [Initialize any needed variables]

  // Step 3: Main algorithm logic
  // [Implement the core algorithm]

  // Step 4: Return result
  return result;
}

/**
 * Optimized version of the algorithm (if applicable)
 *
 * Time Complexity: O(?)
 * Space Complexity: O(?)
 */
function solutionFunctionOptimized(parameters) {
  // Implementation here
}

/**
 * Test cases for the algorithm
 */
function testSolution() {
  console.log('Running tests...');

  // Test case 1: Normal case
  const result1 = solutionFunction([1, 2, 3]);
  console.assert(result1 === expectedResult, 'Test case 1 failed');

  // Test case 2: Edge case - empty input
  const result2 = solutionFunction([]);
  console.assert(result2 === expectedEmptyResult, 'Test case 2 failed');

  // Test case 3: Edge case - single element
  const result3 = solutionFunction([1]);
  console.assert(result3 === expectedSingleResult, 'Test case 3 failed');

  // Test case 4: Edge case - duplicates
  const result4 = solutionFunction([1, 1, 1]);
  console.assert(result4 === expectedDuplicateResult, 'Test case 4 failed');

  console.log('✅ All test cases passed!');
}

// Export for use in other files
module.exports = {
  solutionFunction,
  solutionFunctionOptimized,
  testSolution,
};

// Run tests if this file is executed directly
if (require.main === module) {
  testSolution();
}
